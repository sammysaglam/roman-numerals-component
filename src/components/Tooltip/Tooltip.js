import styled, { css } from 'styled-components';

const beforeAndAfter = css`
	top: 100%;
	left: 30%;
	border: solid transparent;
	content: ' ';
	height: 0;
	width: 0;
	position: absolute;
	pointer-events: none;
`;

const Tooltip = styled.div`
	position: absolute;
	left: -1rem;
	bottom: calc(100% + 1rem);
	background: #88b7d5;
	box-shadow: 1px 2px 10px rgba(0, 0, 0, 0.5);
	padding: 0.375rem 1rem;
	color: white;
	font-weight: bold;
	border: 4px solid #c2e1f5;
	font-family: 'Times New Roman', serif;
	border-radius: 5px;

	&::before {
		${beforeAndAfter};
	}

	&::after {
		${beforeAndAfter};
	}

	&::after {
		border-color: rgba(136, 183, 213, 0);
		border-top-color: #88b7d5;
		border-width: 10px;
		margin-left: -10px;
	}

	&::before {
		border-color: rgba(194, 225, 245, 0);
		border-top-color: #c2e1f5;
		border-width: 16px;
		margin-left: -16px;
	}
`;

export default Tooltip;
