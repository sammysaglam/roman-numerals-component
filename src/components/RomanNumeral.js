import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Tooltip from './Tooltip/Tooltip';
import { toRoman, fromRoman } from '../utils/roman';

const Wrapper = styled.span`
	position: relative;
	cursor: default;
`;

class RomanNumeral extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			isTooltipVisible: false,
		};

		this.onMouseEnter = this.onMouseEnter.bind(this);
		this.onMouseLeave = this.onMouseLeave.bind(this);
	}

	onMouseEnter() {
		this.setState({
			isTooltipVisible: true,
		});
	}

	onMouseLeave() {
		this.setState({
			isTooltipVisible: false,
		});
	}

	render() {
		const { returnType, value, hasTooltip } = this.props;
		const { isTooltipVisible } = this.state;
		const { onMouseEnter, onMouseLeave } = this;

		return (
			<Wrapper onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave}>
				{(returnType === 'roman-numeral' && toRoman(value)) || (returnType === 'decimal' && fromRoman(value))}
				{hasTooltip && isTooltipVisible && <Tooltip>{value}</Tooltip>}
			</Wrapper>
		);
	}
}

RomanNumeral.propTypes = {
	hasTooltip: PropTypes.bool,
	returnType: PropTypes.oneOf(['roman-numeral', 'decimal']),
	value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
};

RomanNumeral.defaultProps = {
	hasTooltip: false,
	returnType: 'roman-numeral',
};

export default RomanNumeral;
