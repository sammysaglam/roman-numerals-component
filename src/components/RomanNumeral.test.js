import React from 'react';
import { shallow } from 'enzyme';

import RomanNumeral from './RomanNumeral';
import Tooltip from './Tooltip/Tooltip';

describe('<RomanNumeral>', () => {
	it('renders Decimal value based on Roman Numeral input', () => {
		const wrapper = shallow(<RomanNumeral value={23423} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders Roman Numeral based on Decimal input', () => {
		const wrapper = shallow(<RomanNumeral returnType="decimal" value="VÌ…DCLXIII" />);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders Tooltip on mouse enter & unmounts Tooltion on mouse leave', () => {
		const wrapper = shallow(<RomanNumeral hasTooltip={true} value={78769} />);

		wrapper.simulate('mouseenter');
		expect(wrapper.find(Tooltip)).toHaveLength(1);

		wrapper.simulate('mouseleave');
		expect(wrapper.find(Tooltip)).toHaveLength(0);
	});

	it('does not render Tooltip on mouse enter if hasTooltip = false', () => {
		const wrapper = shallow(<RomanNumeral hasTooltip={false} value={78769} />);

		wrapper.simulate('mouseenter');

		expect(wrapper.find(Tooltip)).toHaveLength(0);
	});
});
