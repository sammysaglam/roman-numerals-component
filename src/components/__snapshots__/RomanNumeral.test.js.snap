// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RomanNumeral> renders Decimal value based on Roman Numeral input 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RomanNumeral
    hasTooltip={false}
    returnType="roman-numeral"
    value={23423}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        "X̅X̅MMMCDXXIII",
        false,
      ],
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Array [
      "X̅X̅MMMCDXXIII",
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          "X̅X̅MMMCDXXIII",
          false,
        ],
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Array [
        "X̅X̅MMMCDXXIII",
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<RomanNumeral> renders Roman Numeral based on Decimal input 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RomanNumeral
    hasTooltip={false}
    returnType="decimal"
    value="V̅DCLXIII"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        5663,
        false,
      ],
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Array [
      5663,
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          5663,
          false,
        ],
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Array [
        5663,
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
